src/foam/nanos/crunch/connection/ConnectedCapabilityDAO.js

```diff
@@ -51,22 +52,40 @@ foam.CLASS({
         Object[] flatData = sentCap.getData();
 
         // First pass: validate types provided
-        for ( int i = 0 ; i < classes.length ; i++ ) {
-          FObject dataObj = (FObject) flatData[i];
+        int index = 0;
+        for ( var data : flatData ) {
+          // Skip capability where 'of' property is empty
+          while ( index < classes.length && SafetyUtil.isEmpty(classes[index]) ) {
+            index++;
+          }
+
+          if ( index >= classes.length ) {
+            throw new RuntimeException("Unexpected data provided: " + data);
```

Comment:
This seems like a developer/configuration error as well so should generate Log warning/error.  Similar with below.